working/static metaprogramming/intro.md

```diff
@@ -0,0 +1,431 @@
+  FixedStuff fixedStuff;
+  State<Foo> createState() => _FooState();
+}
+
+class _FooState extends State<Foo> {
+  StatefulStuff statefulStuff = ...;
+  @override
+  Widget build(BuildContext context) {
+    buildCode();
+  }
+}
+```
+
+The widget name is repeated six times, `fixedStuff` appears twice, and the
+remaining code is almost pure boilerplate that is identical across all
+`StatefulWidget` and `State` subclasses.
+
+There are other patterns in Flutter like this that are commonly applied to
+widget classes but require Flutter users to write a lot of repetitive code.
+Metaprogramming could enable the Flutter framework to express these patterns
+more tersely and with less chance for user error.
+
+## Requirements
+
+From these sample use cases and others, we can generalize some requirements:
+
+* **Must be able to introspect on the existing program**. Almost all the use
+  cases we want to cover need to be able to look at the fields of an existing
+  class at a minimum, as well as the fields of super classes in many cases. Many
+  also need to inspect the methods, constructors, and parameters of those.
+* **Must be able to synthesize new APIs in existing classes**. Most of the core
+  use cases require adding new methods to classes, whose signatures may not be
+  known ahead of time. Some also require the ability to add new fields, both
+  private and public to classes.
+* **Should be able to synthesize whole new classes**. Some of the core use cases
+  involve creating new classes from either other classes or even just top level
+  functions.
+* **Metaprogramming should be composable**. In one form or another,
+  metaprogramming applications should be able to be composed together. A data
+  class code generator for instance would ideally be a combination of several
+  more targeted code generators for each of the desired features (`copyWith()`,
+  `==`, `hashCode`, etc).
+* **Generated code must be debuggable and user visible**. See the Usability
+  section below.
+
+## Design constraints
+
+In addition to user-level requirements, there are some practical constraints to
+ensure that we don't compromise the existing advantages of Dart.
+
+* **Must support modular compilation**. Any individual library should be able to
+  be fully compiled (including running all macros) using only the transitive
+  dependencies of the library. To this end we should not expose any API or
+  functionality that allows global access to the program, outside of the current
+  library and/or its dependencies. Such functionality would defeat modular
+  compilation.
+* **Must support incremental compilation**. This is key to the hot reload
+  workflow—any change in the program should not have to rerun all
+  metaprogramming in the entire program. It should at worst require a rerun of
+  the metaprogramming in libraries that depend on the modified library.
+* **Must support expression evaluation**. Expression evaluation is a vital tool
```

Comment:
What is "expression evaluation"? Is it evaluating user-input expressions during debugging?

Reply:
Yes